### DISCUSSION TITLE
Handling `text/plain` POST requests as json

### DISCUSSION THREAD

**Discussion by archiif** [fastapi/fastapi#8725] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import uvicorn
from fastapi import FastAPI
from pydantic import BaseModel


class Request(BaseModel):
    param: str


app = FastAPI()


@app.post("/")
def post_request(request: Request):
    pass


uvicorn.run(app, host="localhost", port=9900)
```

### Description

I'm trying to handle requests from an application that is sending json data in the body with `text/plain` as the content type.

This issue seems similar to https://github.com/tiangolo/fastapi/issues/579, but in my case I actually do want to parse the body as json.

Is there a clean way to handle this?

```
C:\>curl -X POST http://localhost:9900 -H "Content-Type: text/plain" -d "{\"param\": \"foobar\"}"
{"detail":[{"loc":["body"],"msg":"value is not a valid dict","type":"type_error.dict"}]
C:\>curl -X POST http://localhost:9900 -H "Content-Type: application/json" -d "{\"param\": \"foobar\"}"
null
```

### Operating System

Windows

### FastAPI Version

0.70.1

### Python Version

Python 3.10.0

**Comment by rafsaf** [comment, marked as answer]
Not sure about "clean" but

```python
import uvicorn
import fastapi
from pydantic import BaseModel
import json

class Request(BaseModel):
    param: str


app = fastapi.FastAPI()


@app.post("/")
async def post_request(fastapi_req: fastapi.Request):
    body = await fastapi_req.body()
    obj = json.loads(body)
    request = Request(**obj)


uvicorn.run(app, host="localhost", port=9900)
```

- Natural next step would be to move this logic somehow (create custom param, header? with parsed body) to [fastapi middleware](https://fastapi.tiangolo.com/tutorial/middleware/) and just add it to endpoints you want. In (first lines of) pointed middleware, check incoming content-type, and if it's "text/plain", try to parse it first and pass forward. Of course this can be done in endpoint also.
- Another idea that can? work is to overwrite in middleware content-type from "text/plain" to "application/json" although there may be some vulnerabilities to investigate, first option with "custom" validation in middleware or endpoints may be better

**Comment by harunyasar** [comment]
Wouldn't it be a solution for you to use the pydantic JSON type? What I mean is that.

```python
from pydantic import BaseModel, Json


class Request(BaseModel):
    param: Json

Request(param='{\"param\": \"foobar\"}')

# Request(param={'param': 'foobar'})
```
